### PR TITLE
Added More Object Detectors

### PR DESCRIPTION
**Pull request by Niro4**
I added RetinaNet and FCOS models to the detection trainer.

### CODE REVIEW

File: .gitignore
Comment (adamjstewart) [inline .gitignore:148]: I'm personally opposed to adding this, but it's borderline. I usually keep a project's `.gitignore` restricted to project-specific things (i.e., programming language) and leave user-specific things (i.e., operating system, text editor) up to global `~/.gitignore` files.

File: torchgeo/trainers/detection.py
Comment (adamjstewart) [inline torchgeo/trainers/detection.py:55]: You can remove this, it isn't needed anymore

**Review by adamjstewart — COMMENTED** [review]
Looks great! A bunch of the tests are failing, so we'll need to get those working. I'll let @nilsleh review things in detail since he recently added timm support to all of our other trainers.

**Comment by Niro4** [comment]
I resolved the requested changes. I'll look into the failed tests. @nilsleh let me know if there is anything I should change or resolutions to some tests of the tests I don't address.

**Comment by adamjstewart** [comment]
https://torchgeo.readthedocs.io/en/stable/user/contributing.html has a bunch of tips for running the tests locally and fixing style issues. Let me know if anything isn't clear!

**Comment by Niro4** [comment]
All tests have passed on my local. The datasets tag was added since there was an error thrown by mypy regarding a comment. I removed the comment, and it seemed to resolve the issue.

**Review by ashnair1 — COMMENTED** [review]
Thanks for the feature. The following comments should help with resolving mypy and minimum tests. The remaining error is caused by a warning raised by `draw_bounding_boxes` (fixed by #988). 

Once #917 and #988 are merged, all tests should pass.

File: torchgeo/datasets/idtrees.py
Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:600]: > The datasets tag was added since there was an error thrown by mypy regarding a comment. I removed the comment, and it seemed to resolve the issue.

This is because you don't have pyvista installed locally. If you install it (like in our CI) then this error will reoccur. Someone needs to contribute type hints to pyvista for this feature so we can remove this ignore. @banesullivan what would the correct type hints for this look like?

Comment (banesullivan) [reply torchgeo/datasets/idtrees.py:600]: I'm still very new to Python type hinting... This class is being directly instantiated so I'm wondering what the confusion is... that class, `pyvista.PolyData`, is the type.

Comment (banesullivan) [reply torchgeo/datasets/idtrees.py:600]: I'll ping others in the PyVista developers list to see if someone more versed in Python type hinting can help here while I read up on this

Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:600]: I would try to investigate this myself but I'm having trouble getting a modern (supports Python 3.10) version of VTK to build: https://github.com/spack/spack/pull/34589. Not sure if any PyVista devs are also VTK devs and could help with that...

Comment (adeak) [reply torchgeo/datasets/idtrees.py:600]: Not that I'm particularly experienced in type hints myself, but this doesn't look like a PyVista type hinting issue. [According to CI](https://github.com/microsoft/torchgeo/actions/runs/3795685078/jobs/6455387965) the mypy error is
>  torchgeo/datasets/idtrees.py:600: error: Module has no attribute "PolyData" [attr-defined]
That's not really a typing error, and definitely not an error that makes any sense.

Looking at the conditional import of `pyvista` in the function, there's another suspicious point:
https://github.com/microsoft/torchgeo/blob/a934fcd62bfccd1c9cd017df31769bf9bb6ff830/torchgeo/datasets/idtrees.py#L586-L587

[According to flake8](https://flake8.pycqa.org/en/2.5.5/warnings.html) F401 means
> `module` imported but unused

So for some reason (due to the conditional import?) the linters think that the `pyvista` name isn't the real PyVista, or that the one imported as `import pyvista` is a different one than what's used later.

Perhaps the solution is to add an
```py
if typing.TYPE_CHECKING:
    import pyvista
```
somewhere? Assuming that plays nicely with typing on systems where pyvista is not available.

Comment (adeak) [reply torchgeo/datasets/idtrees.py:600]: Unless something about `py.typed` (added in https://github.com/pyvista/pyvista/pull/2904) potentially vs. `pyproject.toml` (touched recently) breaks mypy. There are some remarks in the linked PR that could cause mypy not to recognise the _installed_ package, and I think PyVista is tested locally rather than via install. Let me test this on my end.

So this might be a pyvista issue after all, but not the kind I'd imagine, hearing "type hinting" :)

Comment (adeak) [reply torchgeo/datasets/idtrees.py:600]: I checked a module with the following contents:
```py
def foo():
    try:
        import pyvista
    except ImportError:
        raise ImportError()
    
    mesh = pyvista.PolyData()
    return mesh
```
with mypy (0.991, same as your CI) and installed-only PyVista (0.36.1 and 0.37.0 both). I got no error. Any ideas what might be missing from my dummy example? I used `mypy foo.py`.

---

Tangential note: I'd recommend `point_cloud.plot(...)` here: 
https://github.com/microsoft/torchgeo/blob/a934fcd62bfccd1c9cd017df31769bf9bb6ff830/torchgeo/datasets/idtrees.py#L578
;)

Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:600]: Try `mypy --strict foo.py`

Comment (adeak) [reply torchgeo/datasets/idtrees.py:600]: That was it, thanks @adamjstewart, I can repro now.

Comment (adeak) [reply torchgeo/datasets/idtrees.py:600]: I'm starting to hit a wall but let me document what I've figured out so far.

The issue is [due to behaviour change in mypy 0.700](https://github.com/python/mypy/issues/7030). Using `--strict` implies [`--no-implicit-reexport`](https://mypy.readthedocs.io/en/latest/command_line.html#cmdoption-mypy-no-implicit-reexport). What this seems to do is not propagate names imported into modules, including `__init__.py`, so most of the public API of PyVista is not visible to mypy when using `--strict`.

There are a few options I've found so far:
1. Disable the (mis)feature with `mypy --strict --implicit-reexport`: I don't know why the new default of changing typing namespaces to be different from runtime namespaces is preferable by anyone.
2. An official recommendation: use `from foo import bar as bar` everywhere in PyVista (ugh).
3. A slightly better version of the previous point: use `from foo import bar` and explicit `__all__` everywhere in PyVista.

I think point 3 would be preferable anyway in the long run, but PyVista currently has very generous use of star imports in `__init__.py` files. I wouldn't mind pinning (and maybe paring) down the public API, and switching to something explicit via `__all__`. Then again named import _plus_ `__all__` is again redundant. And as it is, enumerating the entire public API in `pyvista/__init__.py` would be... problematic.

I haven't yet figured out if there are other options to circumvent this, e.g. whether the same would happen if we used stub files instead of inline typing. And whether the partialness of out type hints is relevant or not (I suspect not).

Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:600]: Yes, I would also vote for 3, that's [what we do in TorchGeo](https://github.com/microsoft/torchgeo/blob/main/torchgeo/datasets/__init__.py). Star imports are generally [frowned upon](https://www.geeksforgeeks.org/why-import-star-in-python-is-a-bad-idea/), and both mypy and flake8 correctly discourage them. And by discourage I mean they literally don't work if you use star imports.

Comment (adeak) [reply torchgeo/datasets/idtrees.py:600]: They are indeed frowned upon, but `__init__.py` + `__all__` is the one use case where star imports are tolerable. They still make it harder to find where names are from, but namespace bloat and unintended namespace clashes (which are in practice the largest problems with star imports) don't apply there.

But I wouldn't hold it against mypy to break (or just refuse to work) on star inports. However, this issue is more than that: even if we did `from submodule import name, other_name` nicely, mypy would break. The solutions of either using `from submodule import name as name` (bloated, redundant to the point of being obfuscated) or using `__all__` in addition to named imports (which, as far as I know, is a tool to facilitate star imports – ironic?) feel awfully opinionated to me.

Anyway, the way seems clear. Keep the mypy ignore comment here, start discussing import (and public API) cleanup in PyVista.

Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:600]: Thanks for digging into this! Glad we got down to the bottom of this.

**Comment by adamjstewart** [comment]
@Niro4 #917 is close to being merged. Once that's merged, we can rebase this PR on top of that and get it merged. Can you sign the CLA? That's the only thing I can't do for you 😄

**Comment by adamjstewart** [comment]
Update: #917 has been merged

**Comment by Niro4** [comment]
@microsoft-github-policy-service agree [company="SWCA Environmental Consultants"]

**Comment by adamjstewart** [comment]
Need to remove the [ ] brackets I think

**Comment by Niro4** [comment]
@microsoft-github-policy-service agree company="SWCA Environmental Consultants"

**Comment by adamjstewart** [comment]
Next step is to rebase. Not sure how comfortable you are with git but let me know if I can help.

**Comment by Niro4** [comment]
Thanks! I can handle the rebase. It should be updated shortly

File: tests/trainers/test_detection.py
Comment (adamjstewart) [inline tests/trainers/test_detection.py:20]: I didn't even realize you could stack decorators like this to get the full cross product, this is way easier than what I was doing!

File: tests/trainers/test_detection.py
Comment (adamjstewart) [inline tests/trainers/test_detection.py:22]: If we're always testing resnet18, there's no need to parametrize it, we can just use the default in `model_kwargs`. If we had more detection datamodules, I would probably just swap out the module in the config file itself instead of parametrizing, but we can change that in the future.

File: torchgeo/trainers/detection.py
Comment (adamjstewart) [inline torchgeo/trainers/detection.py:48]: Duplicated, we can remove one of these

File: torchgeo/trainers/detection.py
Comment (adamjstewart) [inline torchgeo/trainers/detection.py:86]:
```suggestion
        if self.hyperparams["backbone"] in BACKBONE_LAT_DIM_MAP:
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**
